File: mandelbrot.py
from decimal import *
import random
from tqdm import tqdm

class Mandelbrot:
    def __init__(self, w = Decimal(4), h = Decimal(2), x = Decimal(-1), y = Decimal(0)):
        self.w = w
        self.h = h
        self.x = x
        self.y = y
        self.rendered = None
        self.rendered_res_x = 0
        self.rendered_res_y = 0

    def mandel_point(self, C_x, C_y, iter):
        Z_x = C_x
        Z_y = C_y
        for i in range(0, iter):
            Z_x_old = Z_x
            Z_x = Z_x * Z_x - Z_y * Z_y + C_x
            Z_y = 2 * Z_x_old * Z_y + C_y
            if (Z_x ** 2 + Z_y ** 2) > 4:
                return 1
        return 0

    def render(self, res_x, res_y):
        # Approximation for number of iterations
        iter = int(50 + max(0,-Decimal.log10(self.w)) * 100)

        # Updates precision
        getcontext().prec = int(max(0,-Decimal.log10(self.w))+8)

        columns = []

        for y_offset_i in tqdm(range(res_y, 0, -1)):
            row = []
            for x_offset_i in range(0, res_x):
                p_x = self.x - self.w / Decimal(2) + Decimal(x_offset_i) / Decimal(res_x) * self.w
                p_y = self.y - self.h / Decimal(2) + Decimal(y_offset_i) / Decimal(res_y) * self.h

                row += [self.mandel_point(p_x, p_y, iter)]
            columns += [row]


        self.rendered_res_x = res_x
        self.rendered_res_y = res_y
        self.rendered = columns

    def get_render(self):
        return self.rendered

    def is_area_uniform(self, x_offset, y_offset, w, h, w_div, h_div, w_start, h_start):
        first_point = self.rendered[int(y_offset) + int(h / h_div) * h_start][int(x_offset) + int(w / w_div) * w_start]
        for x in range(0, int(w / w_div)):
            for y in range(0, int(h / h_div)):
                if first_point != self.rendered[int(y_offset) + int(h / h_div) * h_start + y][
                    int(x_offset) + int(w / w_div) * w_start + x]:
                    return False
        return True

    def get_uniformness_of_area(self, w, h, x_offset, y_offset, w_div, h_div):
        uniformness = 0
        for w_start in range(w_div):
            for h_start in range(h_div):
                if self.is_area_uniform(x_offset, y_offset, w, h, w_div, h_div, w_start, h_start):
                    uniformness += 1
        return uniformness

    def zoom_on_interesting_area(self):
        choices = []
        # Upper left
        uniformness = self.get_uniformness_of_area(self.rendered_res_x / 2, self.rendered_res_y / 2, 0, 0, 2, 2)
        choices += [(self.x-self.w/4, self.y+self.h/4, uniformness)]
        # Upper right
        uniformness = self.get_uniformness_of_area(self.rendered_res_x / 2, self.rendered_res_y / 2, self.rendered_res_x / 2, 0, 2, 2)
        choices += [(self.x+self.w/4, self.y+self.h/4, uniformness)]
        # Lower left
        uniformness = self.get_uniformness_of_area(self.rendered_res_x / 2, self.rendered_res_y / 2, 0, self.rendered_res_y / 2, 2, 2)
        choices += [(self.x-self.w/4, self.y-self.h/4, uniformness)]
        # Lower right
        uniformness = self.get_uniformness_of_area(self.rendered_res_x / 2, self.rendered_res_y / 2, self.rendered_res_x / 2, self.rendered_res_y / 2, 2, 2)
        choices += [(self.x + self.w / 4, self.y - self.h / 4, uniformness)]

        self.w = self.w / 2
        self.h = self.h / 2

        # Filter out completely uniform squares
        choices = [x for x in choices if x[2]<4]
        # Filter out squares that have 2 or more uniform squares
        less_uniform_choices = [x for x in choices if x[2]<3]
        if len(less_uniform_choices) != 0:
            self.x, self.y, u = random.choice(less_uniform_choices)
        else:
            self.x, self.y, u = random.choice(choices)

File: mandelbrot.cpp
#include "mandelbrot.hpp"
#include "GUI_Paint.h"
#include <algorithm>
#include <cmath>
#include <vector>
#include <tuple>
#include <iostream>
#include <fstream>
#include <random>

using namespace std;
void MandelbrotSet::InitMandelbrotSet()
{
    w = 4;
    h = 2;
    x = -1;
    y = 0;
    rendered = NULL;
    renderedResX = 0;
    renderedResY = 0;
}

void MandelbrotSet::SetRender(UBYTE* image)
{
    rendered = image;
}

void MandelbrotSet::Render(UWORD xResolution, UWORD yResolution)
{
    // Approximation for number of iterations
    int iter = (50 + max(0.0, -log10(w)) * 100 );
    vector<vector<bool>> columns;

    for(int i = yResolution-1; i>=0; --i)
    {
        vector<bool> rows;
        for(int j = 0; j < xResolution; ++j)
        {
            double p_x = this->x - this->w / 2.0 + (double)j / (double)xResolution * this->w;
            double p_y = this->y - this->h / 2.0 + (double)(i+1) / (double)yResolution * this->h;
            rows.emplace_back(IsMandelPoint(p_x, p_y, iter));
        }
        columns.emplace_back(rows);
    }

    renderedResX = xResolution;
    renderedResY = yResolution;

    // Update rendered image
    for(unsigned int y = 0; y < columns.size(); ++y)
    {
        auto row = columns[y];
        for(unsigned int x = 0; x < row.size(); ++x)
        {
            auto bitSet = row[x];
            if(bitSet)
            {
                Paint_SetPixel(x, y, WHITE);
            }
            else
            {
                Paint_SetPixel(x, y, BLACK);
            }
        }
    }
}

bool MandelbrotSet::IsMandelPoint(double fX, double fY, int iterations)
{
    double z_x = fX;
    double z_y = fY;

    for(int i = 0; i < iterations; ++i)
    {
        double z_x_old = z_x;
        z_x = z_x * z_x - z_y * z_y + fX;
        z_y = 2.0 * z_x_old * z_y + fY;
        auto sumSquared = pow(z_x, 2) + pow(z_y, 2);
        if (sumSquared > 4)
        {
            return true;
        }
    }
    return false;
}

unsigned long long MandelbrotSet::GetUniformnessOfArea(double fW, double fH, int xOffset, int yOffset, int wDiv, int hDiv)
{
    unsigned long long uniformness = 0;
    for(int wStart = 0; wStart < wDiv; ++wStart)
    {
        for(int hStart = 0; hStart < hDiv; ++hStart)
        {
            if(IsAreaUniform(xOffset, yOffset, fW, fH, wDiv, hDiv, wStart, hStart))
            {
                ++uniformness;
            }
        }
    }

    return uniformness;
}

bool MandelbrotSet::IsAreaUniform(int xOffset, int yOffset, double fW, double fH,  int wDiv, int hDiv, double wStart, double hStart)
{
    int yInit = yOffset + static_cast<int>(fH / hDiv) * hStart;
    int xInit = xOffset + static_cast<int>(fW / wDiv) * wStart;
    auto firstPoint = Paint_GetPixel(xInit , yInit);

    for(unsigned int i = 0; i < static_cast<unsigned int>(fW / wDiv); ++i )
    {
        for(unsigned int j = 0; j < static_cast<unsigned int>(fH / hDiv); ++j )
        {
            int yTest = yOffset + static_cast<int>(fH / hDiv) * hStart + j;
            int xTest = xOffset + static_cast<int>(fW / wDiv) * wStart + i;
            auto testPoint = Paint_GetPixel(xTest , yTest);
            if(testPoint != firstPoint)
                return false;
        }        
    }

    return true;
}

void MandelbrotSet::ZoomOnInterestingArea()
{   
    tuple<double, double, unsigned long long> choice;
    vector<tuple<double, double, unsigned long long>> choices;

    auto uniformness = GetUniformnessOfArea(this->renderedResX / 2, this->renderedResY / 2, 0, 0, 2, 2);
    choice = {this->x - this->w/4, this->y + this->h/4, uniformness};

    choices.emplace_back(choice);

    uniformness = GetUniformnessOfArea(this->renderedResX / 2, this->renderedResY / 2, this->renderedResX / 2, 0, 2, 2);
    choice = {this->x + this->w/4, this->y + this->h/4, uniformness};

    choices.emplace_back(choice);

    uniformness = GetUniformnessOfArea(this->renderedResX / 2, this->renderedResY / 2, 0, this->renderedResY / 2, 2, 2);
    choice = {this->x - this->w/4, this->y - this->h/4, uniformness};
    choices.emplace_back(choice);

    uniformness = GetUniformnessOfArea(this->renderedResX / 2, this->renderedResY / 2, this->renderedResX / 2, this->renderedResY / 2, 2, 2);
    choice = {this->x + this->w/4, this->y - this->h/4, uniformness};
    choices.emplace_back(choice);     

    w = w / 2.0;
    h = h / 2.0;

    choices.erase(std::remove_if(
        choices.begin(),
        choices.end(),
        [](const tuple<double, double, unsigned long long>& x) { 
            return (std::get<unsigned long long>(x) >= 4); 
        }), choices.end());

    auto lessUniformChoices = choices;
    lessUniformChoices.erase(std::remove_if(
        lessUniformChoices.begin(),
        lessUniformChoices.end(),
        [](const tuple<double, double, unsigned long long>& x) { 
            return (std::get<unsigned long long>(x) >= 3); 
        }), lessUniformChoices.end());
    
    // Seed
    random_device rd;
    mt19937 g(rd());

    if(lessUniformChoices.size() > 0)
    {
            shuffle(lessUniformChoices.begin(), lessUniformChoices.end(), g);
            auto selection = lessUniformChoices[0];
            this->x = get<0>(selection);
            this->y = get<1>(selection);
    }
    else
    {
            shuffle(choices.begin(), choices.end(), g);
            auto selection = choices[0];
            this->x = get<0>(selection);
            this->y = get<1>(selection);
    }
}

